src/runtime_src/xocl/api/clReleaseMemObject.cpp

```diff
@@ -41,6 +41,10 @@ clReleaseMemObject(cl_mem memobj)
 {
   validOrError(memobj);
 
+  //Cleanup any connection info before decrementing
+  //the reference count and then perhaps proceeding to deletion.
+  xocl(memobj)->cleanup();
```

Comment:
Cleanup analogous to decrementing of the ref count ; similarly if there is a connection saved for it, it should be deleted. I did consider putting it in free(mem) call in xocl/core/device.cpp but then realised that it needs to go before we return if refcount !=0.

PS: tested this extensively with all types of buffer object creations, mutplle, using ext_ptr_t the old way, the new way and with XCL_MEM_TOPOLOGY.

Reply:
I don't understand?  Connections are per memory object, not per refcount?  I am confused.

Reply:
will show you via desktop sharing the issue i was having if i placed it after release() call.

Reply:
I can understand there is a problem when memory objects are released indirectly after the program and kernels have been deleted.  If the program is deleted so is the xclbin. But clearing the index  in clReleaseMemObject is no guarantee that it will not be after program is gone.  It's all a matter of the order in which user calls clRelease[Objects].